### PR TITLE
Feature/dashboard teacher/bulk signature

### PR DESCRIPTION
**Pull request by jbpenrath**
## Purpose

The training contract signature has evolved and no need that organization can sign a contract. So this PR updates the TeacherDashboard and api interfaces to add this feature.

https://github.com/openfun/richie/assets/9265241/b2f7cb4e-6bc5-47b6-a104-669a02e0ab29

## Proposal

- [x] Update TeacherDashboard views to sign all pending contracts.

### CODE REVIEW

File: src/frontend/js/pages/TeacherDashboardContractsLoader/ContractFilters/index.tsx [hunk 1]
```diff
@@ -1,43 +1,123 @@
...
+export const messages = defineMessages({
   organizationFilterLabel: {
     defaultMessage: 'Organization',
     description: 'Use as organization filter label',
     id: 'pages.TeacherDashboardContractsLoader.ContractFilters.organizationFilterLabel',
   },
+  contractStateFilterLabel: {
+    defaultMessage: 'Signature state',
+    description: 'Contract signature state filter label',
+    id: 'pages.TeacherDashboardContractsLoader.ContractFilters.contractSignatureStateFilterLabel',
```
Comment (rlecellier) [inline src/frontend/js/pages/TeacherDashboardContractsLoader/ContractFilters/index.tsx:18]:
```suggestion
    'pages.TeacherDashboardContractsLoader.ContractFilters.contractStateFilterLabel',
```

File: src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.tsx [hunk 2]
```diff
@@ -1,144 +1,184 @@
...
+  }
 
-  return contracts ? (
-    <div className="teacher-contract-page" data-testid={isFetched ? 'contracts-loaded' : undefined}>
+  return (
+    <div className="teacher-contract-page">
+      <div className="dashboard__page__actions">
+        <div className="dashboard__page__actions-row dashboard__page__actions-row--space-between">
+          <div>
+            {filters.organization_id && canSignContract && (
```
Comment (rlecellier) [inline src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.tsx:124]: i would have directly use `contractAbilities.can(ContractActions.SIGN);` here

File: src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.tsx [hunk 3]
```diff
@@ -1,144 +1,184 @@
...
-    return contracts.map(({ id, student_signed_on: studentSignedOn, order }) => ({
-      id,
-      learnerName: order.owner_name,
-      productTitle: order.product_title,
-      signDate: studentSignedOn ? intl.formatDate(studentSignedOn) : studentSignedOn,
-    }));
-  }, [contracts]);
+  if (error) {
+    return <Banner message={error} type={BannerType.ERROR} rounded />;
+  }
```
Comment (rlecellier) [inline src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.tsx:117]: This may replace the datagrid.
if a request fail on a organization, i don't want the page layout to disapear.

File: src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.tsx [hunk 4]
```diff
@@ -1,144 +1,184 @@
...
+                      values={{ count: contractToSignCount }}
+                    />
+                  </Button>
+                )}
+                <OrganizationContractFrame
+                  organizationId={filters.organization_id}
+                  isOpen={contractFrameOpened}
+                  onClose={() => setContractFrameOpened(false)}
+                />
+              </>
```
Comment (rlecellier) [inline src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.tsx:149]: I would have made a `SignOrganizationContractButton` component.

File: src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardCourseContractsLoader/index.tsx [hunk 5]
```diff
@@ -1,50 +1,26 @@
...
-  },
 });
 
 export const TeacherDashboardCourseContractsLoader = () => {
-  const { courseId } = useParams<{ courseId?: string }>();
-  const { page = '1' }: { page?: string } = queryString.parse(location.search);
-  const [filters, setFilters] = useState<ContractListFilters>({});
-  const onFiltersChange = (newFilters: ContractListFilters) => {
-    setFilters(newFilters);
-  };
```
Comment (rlecellier) [inline src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardCourseContractsLoader/index.tsx:31]: The idea of these `*Loader` is to handle outside world interaction that will initialize the page.
It handle the initial loading.
If you remove this logic from here, you can remove the component.

File: src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.spec.tsx [hunk 6]
```diff
@@ -20,129 +23,273 @@ jest.mock('utils/context', () => ({
...
+    );
+
+    render(
+      <Wrapper
+        path="/courses/:courseId/products/:productId/contracts"
+        initialEntry="/courses/1/products/2/contracts"
+      />,
+    );
+
+    await waitFor(expectNoSpinner);
```
Comment (rlecellier) [inline src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.spec.tsx:98]:
```suggestion
    await expectNoSpinner();
```

File: src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.spec.tsx [hunk 7]
```diff
@@ -20,129 +23,273 @@ jest.mock('utils/context', () => ({
...
+
+    // Table listing all signed contracts should have been rendered
+    screen.getByRole('table');
+
+    // Table header should have been rendered with 3 columns
+    const columnHeaders = screen.getAllByRole('columnheader');
+    expect(columnHeaders.length).toBe(3);
+    expect(columnHeaders[0]).toHaveTextContent('Training');
+    expect(columnHeaders[1]).toHaveTextContent('Learner');
+    expect(columnHeaders[2]).toHaveTextContent('State');
```
Comment (rlecellier) [inline src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.spec.tsx:185]: no need to test header twice.

File: src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.spec.tsx [hunk 8]
```diff
@@ -20,129 +23,273 @@ jest.mock('utils/context', () => ({
...
+
+    // Table body should have been rendered with 3 rows (one per contract)
+    contracts.forEach((contract) => {
+      const row = screen.getByTestId(contract.id);
+      const cells = getAllByRole(row, 'cell');
+      expect(cells.length).toBe(3);
+      expect(cells[0]).toHaveTextContent(contract.order.product_title);
+      expect(cells[1]).toHaveTextContent(contract.order.owner_name);
+      expect(cells[2]).toHaveTextContent('Pending for signature');
```
Comment (rlecellier) [inline src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.spec.tsx:254]:
```suggestion
      expect(cells[2]).toHaveTextContent('Pending for signature');
```

**Review by rlecellier — COMMENTED** [review]
Lot's of changes!
I would not have remove loading logic from `ContractLoader`, it would have been nice to split test between:
* listing rendering: headers and row contents and update when contracts prop change.
* router interaction: filter initialization, api call on filter change

File: src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.spec.tsx [hunk 9]
```diff
@@ -20,129 +23,273 @@ jest.mock('utils/context', () => ({
...
+              </Routes>
+            </MemoryRouter>
+          </CunninghamProvider>
         </JoanieSessionProvider>
       </QueryClientProvider>
     </IntlProvider>
   );
 };
 
 describe('pages/TeacherDashboardContracts', () => {
```
Comment (NathanVss) [inline src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.spec.tsx:55]: we could maybe add a test for empty list

File: src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.tsx [hunk 10]
```diff
@@ -1,144 +1,184 @@
...
+    <div className="teacher-contract-page">
+      <div className="dashboard__page__actions">
+        <div className="dashboard__page__actions-row dashboard__page__actions-row--space-between">
+          <div>
+            {filters.organization_id && canSignContract && (
+              <>
+                {hasContractToSign && (
+                  <Button
+                    size="small"
+                    color="secondary"
```
Comment (NathanVss) [inline src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.tsx:129]: Semantic: it seems to be the main action of this page, so it may need to be primary? WDYT?

File: src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.tsx (hunk 4 above)
Comment (NathanVss) [reply src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.tsx:149]: Agreed

Comment (NathanVss) [reply src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.tsx:149]: Capture d’écran 2023-12-21 à 15 35 54

**Review by NathanVss — APPROVED** [review]
## Awesome work! 👏


**Small bug: the contracts from cancelled orders still appear as pending signature:**

Capture d’écran 2023-12-21 à 12 04 49

---

**WDYT about having the button on the same floating line of filter ? On responsive mode it should wrap of course**

Capture d’écran 2023-12-21 à 15 37 34

File: src/frontend/js/api/joanie.spec.ts [hunk 11]
```diff
@@ -55,4 +55,20 @@ describe('api/joanie', () => {
...
+    };
+    const url = buildApiUrl(
+      'http://example.com/organizations/:organization_id/contracts/:id/',
+      filters,
+    );
+
+    expect(url).toBe(
+      'http://example.com/organizations/org_1/contracts/?id=1&id=2&id=3&state=half_signed',
+    );
+  });
```
[the remark this answers is not in the document]
Comment (jbpenrath) [reply src/frontend/js/api/joanie.spec.ts:73]: Yup.

We are using same route to list and retrieve resource.
e.g `organizations/:id/`
The buildApiUrl must be able to bind id to the url when we want to get a resource through its id but also clean up the parameter when it is not provided to list the resource.

**Comment by jbpenrath** [comment]
> ## Awesome work! 👏
> **Small bug: the contracts from cancelled orders still appear as pending signature:**
> 
> Capture d’écran 2023-12-21 à 12 04 49

This has been fixed in openfun/joanie 🙏 

> **WDYT about having the button on the same floating line of filter ? On responsive mode it should wrap of course**
> 
> Capture d’écran 2023-12-21 à 15 37 34

I'll go with that first. A bulk download button will soon appear and on course views, a select to switch from organization is also displayed. That's why I used this row layout.

Many thanks !

File: src/frontend/js/api/joanie.ts [hunk 12]
```diff
@@ -164,6 +164,10 @@ export const getRoutes = () => {
       courses: {
         get: `${baseUrl}/organizations/:organization_id/courses/:id/`,
       },
+      contracts: {
+        get: `${baseUrl}/organizations/:organization_id/contracts/:id/`,
```
[the remark this answers is not in the document]
Comment (jbpenrath) [reply src/frontend/js/api/joanie.ts:168]: This route is used under the hood within API.contracts.get

File: src/frontend/js/widgets/Dashboard/components/DashboardSidebar/_styles.scss [hunk 13]
```diff
@@ -66,18 +66,34 @@
         display: none;
       }
 
-      li {
-        height: rem-calc(16px);
-        overflow: hidden;
+      &__item {
```
[the remark this answers is not in the document]
Comment (jbpenrath) [reply src/frontend/js/widgets/Dashboard/components/DashboardSidebar/_styles.scss:69]: Indeed, I was in rush and I did not take time to extract those part...

File: src/frontend/js/widgets/Dashboard/components/DashboardSidebar/index.tsx [hunk 14]
```diff
@@ -48,11 +55,16 @@ export const DashboardSidebar = ({
...
-              className={selectedLink && selectedLink === link.to ? 'active' : undefined}
+              className={classNames('dashboard-sidebar__container__nav__item', {
+                active: selectedLink === link.to,
+              })}
             >
               <NavLink to={link.to} end>
                 {link.label}
               </NavLink>
+              {link.badge && (
+                <div className="sidebar__container__nav__item__bullet">{link.badge}</div>
```
[the remark this answers is not in the document]
Comment (jbpenrath) [reply src/frontend/js/widgets/Dashboard/components/DashboardSidebar/index.tsx:68]: Good catch dev residue

File: src/frontend/js/settings.ts [hunk 15]
```diff
@@ -39,7 +39,7 @@ export const CONTRACT_SETTINGS = {
   // Interval in ms to poll the related order when a signature has succeeded.
   pollInterval: 1000,
   // Number of retries
-  pollLimit: 30,
+  pollLimit: 45,
```
[the remark this answers is not in the document]
Comment (jbpenrath) [reply src/frontend/js/settings.ts:42]: Indeed good idea !

File: src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.tsx (hunk 10 above)
Comment (jbpenrath) [reply src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.tsx:129]: Yup indeed !

File: src/frontend/js/api/joanie.ts [hunk 16]
```diff
@@ -364,18 +368,14 @@ const API = (): Joanie.API => {
...
-          else url = ROUTES.user.contracts.get.replace(':id/', '');
-
-          if (!ObjectHelper.isEmpty(queryParameters)) {
-            url += '?' + queryString.stringify(queryParameters);
-          }
-
-          return fetchWithJWT(url, { method: 'GET' }).then(checkStatus);
+        get: async ({ contract_ids, ...filters } = {}) => {
+          const endpointFilters = { ...filters, queryParameters: { id: contract_ids } };
+          return fetchWithJWT(
```
[the remark this answers is not in the document]
Comment (jbpenrath) [reply src/frontend/js/api/joanie.ts:373]: Yup but we move this logic into the `useContracts` hooks

File: src/frontend/js/utils/AbilitiesHelper/index.ts [hunk 17]
```diff
@@ -1,29 +1,47 @@
...
-    return !!joanieUserProfileAbilities[action](entity);
-  }
+  if (!Array.isArray(entities)) entities = [entities];
+
+  return entities.every((entity) => {
+    if (isJoanieUserProfileEntity(entity)) {
+      return joanieUserProfileAbilities[action as JoanieUserProfileActions](entity);
+    }
+
+    if (isContractEntity(entity)) {
```
[the remark this answers is not in the document]
Comment (jbpenrath) [reply src/frontend/js/utils/AbilitiesHelper/index.ts:36]: Yup and also the open close principle.

File: src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.tsx (hunk 3 above)
Comment (jbpenrath) [reply src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.tsx:109]: Oh yes right !

File: src/frontend/js/api/joanie.ts [hunk 18]
```diff
@@ -364,18 +368,14 @@ const API = (): Joanie.API => {
...
-          if (id) url = ROUTES.user.contracts.get.replace(':id', id);
-          else url = ROUTES.user.contracts.get.replace(':id/', '');
-
-          if (!ObjectHelper.isEmpty(queryParameters)) {
-            url += '?' + queryString.stringify(queryParameters);
-          }
-
-          return fetchWithJWT(url, { method: 'GET' }).then(checkStatus);
+        get: async ({ contract_ids, ...filters } = {}) => {
+          const endpointFilters = { ...filters, queryParameters: { id: contract_ids } };
```
[the remark this answers is not in the document]
Comment (jbpenrath) [reply src/frontend/js/api/joanie.ts:372]: After a quick check, the url max length limit should not be a problem. The main limitation is related to the SEO and browser, this is not the case here.

File: src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.tsx (hunk 3 above)
Comment (jbpenrath) [reply src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.tsx:109]: Only the datagrid and action buttons will disappear.

Capture d’écran 2024-01-02 à 11 48 09

File: src/frontend/js/components/ContractFrame/AbstractContractFrame.tsx [hunk 19]
```diff
@@ -86,53 +84,84 @@ enum ContractSteps {
...
 }
 
+export interface AbstractProps {
+  isOpen: boolean;
+  onDone?: () => void;
+  onClose?: () => void;
+}
+
+export interface Props extends AbstractProps {
+  getInvitationLink: () => Promise<string>;
```
[the remark this answers is not in the document]
Comment (jbpenrath) [reply src/frontend/js/components/ContractFrame/AbstractContractFrame.tsx:94]: As a Promise is returned, it implies that the function is async

File: src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardCourseContractsLoader/index.tsx (hunk 5 above)
Comment (jbpenrath) [reply src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardCourseContractsLoader/index.tsx:31]: Indeed, the term Loader is no more relevant for this kind of component, I should replace Loader by Layout.

File: src/frontend/js/utils/AbilitiesHelper/index.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (jbpenrath) [reply src/frontend/js/utils/AbilitiesHelper/index.ts:36]: There is something to refactor here, but I don't think we should do it in this PR.

File: src/frontend/js/components/ContractFrame/AbstractContractFrame.tsx [hunk 20]
```diff
@@ -86,53 +84,84 @@ enum ContractSteps {
...
-  const queryClient = useQueryClient();
   const timeoutRef = useRef<NodeJS.Timeout>();
 
   const setErrored = (e: string) => {
     setStep(ContractSteps.ERROR);
     setError(e);
   };
 
-  const initialize = async () => {
+  const onSigned = () => {
```
[the remark this answers is not in the document]
Comment (jbpenrath) [reply src/frontend/js/components/ContractFrame/AbstractContractFrame.tsx:137]: Not quit sure to understand your comment. I do not rename `initialize` method by `onSigned`, I just moved up the `onSigned` method and move down the `initialize` that has been renamed `start`

File: src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.spec.tsx (hunk 9 above)
Comment (jbpenrath) [reply src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.spec.tsx:55]: Right !

File: src/frontend/js/api/joanie.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/api/joanie.ts:373]: `useOrganizationsContracts`, `useUsersContracts`? Looks overkill 😅 
I think we can live with the current code but it makes me think that maybe we should have a standalone `/api/contracts` route that can handle various query params such as `organization_id` etc ..

File: src/frontend/js/utils/AbilitiesHelper/index.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/utils/AbilitiesHelper/index.ts:36]: ok deal :)

**Review by NathanVss — APPROVED** [review]
Let's 🛳️ it!! Well done 👏
